### PR TITLE
Fully deprecate TIMEOUT and NO_RELOAD

### PR DESCRIPTION
**Pull request by ihac**
### 1. Why is this pull request needed and what does it do?
Since the deprecation of `TIMEOUT` and `NO_RELOAD` was announced in v1.4.0, this PR fully deprecates them and refactors related unit tests.

### 2. Which issues (if any) are related?
[#2537: plugin/auto|file: fully deprecate TIMEOUT + NO_RELOAD](https://github.com/coredns/coredns/issues/2537)

### 3. Which documentation changes (if any) need to be made?
This PR removes the docs related to `TIMEOUT` and `NO_RELOAD`.

### 4. Does this introduce a backward incompatible change or deprecation?
Yes, old configuration with `TIMEOUT` or `NO_RELOAD` will not be parsed.

### CODE REVIEW

**Comment by ihac** [comment]
Hi, @stp-ip @miekg 
This PR deprecates the use of `TIMEOUT` and `NO_RELOAD` fully as you planned in [issue 2537](https://github.com/coredns/coredns/issues/2537). However, according to the deprecation policy, CoreDNS is expected to be compatible with old configuration in `Release x.y+1.0` (i.e. 1.5.0). My PR seems to break the rule because all old Corefiles with `TIMEOUT` or `NO_RELOAD` used will not work. This is the reason why CI failed.
I'm a little bit confused~Do I need to update the CI tests?

**Comment by miekg** [comment]
We're jumping to 150 from 140, I honestly can't remember what's now being deprecated other than this was on the list at some point. So I'll just merged when this is not failing CI.

**Comment by miekg** [comment]
Yes tests need to be updated as well and some readme is now failing to parse:
~~~
    metrics_test.go:88: Could not get CoreDNS serving instance: plugin/auto: Corefile:3 - Error during parsing: Wrong argument count or unexpected line ending after '1'
=== RUN   TestMetricsSeveralBlocs
--- PASS: TestMetricsSeveralBlocs (0.02s)
=== RUN   TestMetricsPluginEnabled
--- PASS: TestMetricsPluginEnabled (0.00s)
=== RUN   TestLookupBalanceRewriteCacheDnssec
=== PAUSE TestLookupBalanceRewriteCacheDnssec
=== RUN   TestProxyThreeWay
--- PASS: TestProxyThreeWay (6.00s)
=== RUN   TestLookupProxy
=== PAUSE TestLookupProxy
=== RUN   TestReadme
--- FAIL: TestReadme (6.05s)
    readme_test.go:72: Failed to start server with ../plugin/auto/README.md, for input "plugin/auto: Corefile:3 - Error during parsing: Wrong argument count or unexpected line ending after '45'":
        org {
            auto {
                directory /etc/coredns/zones/org www\.db\.(.*) {1} 45
            }
        }
~~~

**Comment by ihac** [comment]
Thanks @miekg 
Tests and readme now have been updated.

File: plugin/auto/README.md [hunk 1]
```diff
@@ -15,9 +15,8 @@ zonefile. New or changed zones are automatically picked up from disk.
 
 ~~~
 auto [ZONES...] {
-    directory DIR [REGEXP ORIGIN_TEMPLATE [TIMEOUT]]
+    directory DIR [REGEXP [ORIGIN_TEMPLATE]]
```
Comment (miekg) [inline plugin/auto/README.md]: you've added a `[` here that wasn't here before.

Comment (miekg) [reply plugin/auto/README.md]: should it be here?

**Review by miekg — COMMENTED** [review]
lgtm

1 question

File: plugin/auto/README.md (hunk 1 above)
Comment (ihac) [reply plugin/auto/README.md]: Actually I was trying to make it stay consistent with the code, where `TEMPLATE` is optional. 

``` golang
			case "directory": // directory DIR [REGEXP [TEMPLATE]]
				if !c.NextArg() {
					return a, c.ArgErr()
				}
				a.loader.directory = c.Val()
                                 // ...

				// regexp
				if c.NextArg() {
                                         // ...
			        }
				// template is optional here
				if c.NextArg() {
					a.loader.template = rewriteToExpand(c.Val())
				}
```

I'm not sure whether you were making `TEMPLATE` optional on purpose. I think it is fine, but we should set a default value (e.g. `{0}`). WDYT?
BTW, I personally prefer a `return fast` coding style (return as soon as no args can be parsed).

**Comment by stp-ip** [comment]
Partially the warning was already present and the docs were already modified before 1.4.0, but we didn't cut a minor release and so these changes were officially announced with 1.4.0.
In my view deprecating them in 1.5.0 seems ok. Also it's not the main breaking change.
Also taking a look over the code in a sec.

**Review by stp-ip — APPROVED** [review]
/lgtm

Small nitpick related to the plugin/auto docs:
We show "transfer to/from" within the example, but don't actually document the variable etc.

I would add this from plugin/file:
```
transfer enables zone transfers. It may be specified multiples times. To or from signals the direction. ADDRESS must be denoted in CIDR notation (e.g., 127.0.0.1/32) or just as plain addresses. The special wildcard * means: the entire internet (only valid for 'transfer to'). When an address is specified a notify message will be send whenever the zone is reloaded.
```

File: plugin/auto/README.md (hunk 1 above)
Comment (stp-ip) [reply plugin/auto/README.md]: As far as my understanding of intent goes:
The optionality and the default is for both the regex and the template as they are quite correlated. That's why it wasn't separately shown as optional, but one "bundle".

`directory DIR [REGEXP ORIGIN_TEMPLATE]`

The code should reflect the default for this bundle to be `db\.(.*) {1}` afaik.

**Comment by ihac** [comment]
@stp-ip Oops, I just forgot it~~
I'll include it in this PR.

File: plugin/auto/README.md (hunk 1 above)
Comment (ihac) [reply plugin/auto/README.md]: That make sense. So we should raise an error when `REGEXP` is set but `ORIGIN_TEMPLATE` is not?

**Comment by stp-ip** [comment]
> @stp-ip Oops, I just forgot it~~
> I'll include it in this PR.

That sounds reasonable.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Fully depreca..." ]
>ihac commented on this pull request.
>
>
>
>> @@ -15,9 +15,8 @@ zonefile. New or changed zones are automatically picked up from disk.
>
> ~~~
> auto [ZONES...] {
>-    directory DIR [REGEXP ORIGIN_TEMPLATE [TIMEOUT]]
>+    directory DIR [REGEXP [ORIGIN_TEMPLATE]]
>
>That make sense. So we should raise an error when `REGEXP` is set but `ORIGIN_TEMPLATE` is not?

That can wait (a bit) and should probably be done in a new PR (as it's not 
really related to this one)

**Comment by ihac** [comment]
I've updated the PR (remove redundant `[`).

**Pull request merged**